################################################################################
SHORT DESCRIPTION: 
################################################################################
Remove PNP_DISABLE capability for serial port on 0x3f8 as Xen uses it and will
disable it when going to S3.

################################################################################
LONG DESCRIPTION: 
################################################################################
Disables the functionality to evaluate ACPI _DIS method on serial 0x3f8 which
is used by Xen logging. Fixes S3 for machines with PNP serial port. Xen is
supposed to disable the serial port as it uses it.

################################################################################
CHANGELOG 
################################################################################
Original Author: unknown
Port to 3.18: Eric Chanudet <chanudete@ainfosec.com>

################################################################################
REMOVAL 
################################################################################
Without this work-around, S3 is known to fail on machines with PNP serial ports.
(Optiplex 980 for example).

################################################################################
UPSTREAM PLAN
################################################################################
None.

################################################################################
INTERNAL DEPENDENCIES 
################################################################################
None.

################################################################################
PATCHES 
################################################################################
Index: linux-4.9.6/drivers/tty/serial/8250/8250_pnp.c
===================================================================
--- linux-4.9.6.orig/drivers/tty/serial/8250/8250_pnp.c
+++ linux-4.9.6/drivers/tty/serial/8250/8250_pnp.c
@@ -457,6 +457,10 @@ serial_pnp_probe(struct pnp_dev *dev, co
 	} else if (pnp_port_valid(dev, 0)) {
 		uart.port.iobase = pnp_port_start(dev, 0);
 		uart.port.iotype = UPIO_PORT;
+               /* Xen uses port 3f8 for logging, turn off ability to disable it as it breaks suspend */
+               if (uart.port.iobase == 0x3f8) {
+                       dev->capabilities &= ~PNP_DISABLE;
+               }
 	} else if (pnp_mem_valid(dev, 0)) {
 		uart.port.mapbase = pnp_mem_start(dev, 0);
 		uart.port.iotype = UPIO_MEM;
